include/atomics/sender_cadmium.hpp

```diff
@@ -0,0 +1,186 @@
+
+        if((get_messages<typename defs::controlIn>(mbs).size()+
+            get_messages<typename defs::ackIn>(mbs).size())>1){
+            assert(false && "one message per time uniti");
+        }
+        for(const auto &x : get_messages<typename defs::controlIn>(mbs)){
+            if(state.model_active == false){
+                state.total_packet_num = static_cast < int > (x.value);
+                if (state.total_packet_num > 0){
+                    state.packet_num = 1;
+                    state.ack = false;
+                    state.sending = true;
+                    state.alt_bit = state.packet_num % 2;  //set initial alt_bit
+                    state.model_active = true;
+                    state.next_internal = PREPARATION_TIME;
+                }
+                else{
+                    if(state.next_internal != std::numeric_limits<TIME>::infinity()){
+                        state.next_internal = state.next_internal - e;
+                    }
+               }
+            }
+        }
+        for(const auto &x : get_messages<typename defs::ackIn>(mbs)){
+            if(state.model_active == true){ 
+                if (state.alt_bit == static_cast < int > (x.value)){
+                    state.ack = true;
+                    state.sending = false;
+                    state.next_internal = TIME("00:00:00");
+                }
+                else{
+                    if(state.next_internal != std::numeric_limits<TIME>::infinity()){
+                        state.next_internal = state.next_internal - e;
+                    }
+                }
+            }
+        }  
+                           
+    }
+
+    // confluence transition
+    void confluence_transition(TIME e, 
+                               typename make_message_bags<input_ports>::type mbs){
+        internal_transition();
+        external_transition(TIME(), std::move(mbs));
+    }
+
+    // output function
+    typename make_message_bags<output_ports>::type output() const {
+        typename make_message_bags<output_ports>::type bags;
+        message_t out;
+        if (state.sending){
+            out.value = state.packet_num * 10 + state.alt_bit;
+            get_messages<typename defs::dataOut>(bags).push_back(out);
+            out.value = state.packet_num;
+            get_messages<typename defs::packetSentOut>(bags).push_back(out);
+        }
+        else{
+            if(state.ack){
+                out.value = state.alt_bit;
+                get_messages<typename defs::ackReceivedOut>(bags).push_back(out);
+            }
+        }   
+        return bags;
+    }
+
+    // time_advance function
+    TIME time_advance() const {  
+        return state.next_internal;
+    }
+
+    friend std::ostringstream& operator<<(std::ostringstream& os, 
+                                          const typename Sender<TIME>::state_type& i){
+        os << "packet_num: " << i.packet_num << " & total_packet_num: " << i.total_packet_num; 

```

Comment:
It is 95 in length and can be made short by  breaking into two lines

Reply:
resolved as per suggestion